core/src/main/scala/io/projectglow/sql/expressions/LogisticRegressionGwas.scala

```diff
@@ -173,48 +168,169 @@ class NewtonArguments(
 }
 
 case class NewtonResult(
-    args: NewtonArguments,
+    args: NewtonIterationsState,
     logLkhd: Double,
     nIter: Int,
     converged: Boolean,
     exploded: Boolean)
 
 trait LogitTest extends Serializable {
+  type FitState
   def resultSchema: StructType
+  def canReuseNullFit: Boolean
+  def fitNullModel(phenotypes: Array[Double], covariates: SparkDenseMatrix): FitState
+  def runTest(
+      x: DenseMatrix[Double],
+      y: DenseVector[Double],
+      nullModelFit: Option[FitState]): InternalRow
 }
 
-trait LogitTestWithNullModelFit extends LogitTest {
-  def runTest(nullFit: NewtonResult): InternalRow
+trait LogitTestWithoutNullFit extends LogitTest {
+  final override val canReuseNullFit: Boolean = false
+
+  final override def fitNullModel(
+      phenotypes: Array[Double],
+      covariates: SparkDenseMatrix): FitState = {
+    throw new IllegalArgumentException("Test does not support null model fit")
```

Comment:
Can we add a test to make sure we don't call this on accident?

Reply:
Not sure what you mean here. Since we call both tests from Spark, we do verify that this method is never called on an existing test.